Simplify Navigation::scriptExecutionContext()

TingPing opened the pull request:
#### 30c07bfcf389ac842ae25fd40d99c9b1913d78af
<pre>
Simplify Navigation::scriptExecutionContext()
<a href="https://bugs.webkit.org/show_bug.cgi?id=271161">https://bugs.webkit.org/show_bug.cgi?id=271161</a>

Reviewed by Chris Dumez.

As Navigation is a LocalDOMWindowProperty it always has
access to its own ScriptExecutionContext.

* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::navigation):
* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::Navigation):
(WebCore::Navigation::initializeEntries):
(WebCore::Navigation::scriptExecutionContext const):
(WebCore::Navigation::serializeState):
(WebCore::Navigation::navigate):
(WebCore::Navigation::updateCurrentEntry):
(WebCore::Navigation::updateForNavigation):
* Source/WebCore/page/Navigation.h:
* Source/WebCore/page/Navigation.idl:

Canonical link: <a href="https://commits.webkit.org/276478@main">https://commits.webkit.org/276478@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6b35672d9ffa5cf5ebae1bb63cb5e01b02c39949

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/44745 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/23838 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/47227 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/47399 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/40750 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/47048 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/27923 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/21231 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/36782 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/45323 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/20952 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/38553 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/17847 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/18375 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/39687 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/2792 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/41026 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/39976 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/49060 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/19709 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/16278 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/43759 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/21030 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/42506 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/21367 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6197 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/20704 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rwlbuis reviewed (commented):
LGTM.

cdumez commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -178,9 +178,9 @@ Navigation::Result Navigation::reload(ReloadOptions&& options, Ref<DeferredPromi
+Navigation::Result Navigation::navigate(const String& url, NavigateOptions&& options, Ref<DeferredPromise>&& committed, Ref<DeferredPromise>&& finished)
 {
-    auto currentURL = scriptExecutionContext.url();
+    auto currentURL = scriptExecutionContext()->url();
```
The code change looks right to me. To get the resolve the URL, you're using the *relevant* script execution context (instead of the *current* one), which is what the spec [1] says:
> Let urlRecord be the result of encoding-parsing a URL given url, relative to this's relevant settings object.

[1] https://html.spec.whatwg.org/multipage/nav-history-apis.html#dom-navigation-navigate

cdumez commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -76,7 +75,7 @@ bool Navigation::canGoForward() const
 {
     for (Ref item : items)
-        m_entries.append(NavigationHistoryEntry::create(protectedScriptExecutionContext().get(), item));
+        m_entries.append(NavigationHistoryEntry::create(scriptExecutionContext(), item));
```
This is a regression, we should be passing a smart pointer here, please call `protectedScriptExecutionContext().get()`. If `protectedScriptExecutionContext()` is missing, please add it.

cdumez commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -135,7 +135,7 @@ ExceptionOr<RefPtr<SerializedScriptValue>> Navigation::serializeState(JSC::JSVal
 
     Vector<RefPtr<MessagePort>> dummyPorts;
-    auto serializeResult = SerializedScriptValue::create(*protectedScriptExecutionContext()->globalObject(), state, { }, dummyPorts, SerializationForStorage::Yes);
+    auto serializeResult = SerializedScriptValue::create(*scriptExecutionContext()->globalObject(), state, { }, dummyPorts, SerializationForStorage::Yes);
```
Please keep the original code. Why revert our smart pointer adoption work?

cdumez commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -366,7 +366,7 @@ void Navigation::updateForNavigation(Ref<HistoryItem>&& item, FrameLoadType fram
 
     if (navigationType == NavigationNavigationType::Push || navigationType == NavigationNavigationType::Replace)
-        m_entries[*m_currentEntryIndex] = NavigationHistoryEntry::create(protectedScriptExecutionContext().get(), item);
+        m_entries[*m_currentEntryIndex] = NavigationHistoryEntry::create(scriptExecutionContext(), item);
```
Please keep original code, this reverts our smart pointer adoption work.

cdumez commented on Source/WebCore/page/Navigation.h:
```diff
@@ -115,20 +115,20 @@ class Navigation final : public RefCounted<Navigation>, public EventTarget, publ
 
 private:
-    Navigation(ScriptExecutionContext*, LocalDOMWindow&);
+    Navigation(LocalDOMWindow&);
```
Should be marked as `explicit`

TingPing committed:
Simplify Navigation::scriptExecutionContext()
https://bugs.webkit.org/show_bug.cgi?id=271161

Reviewed by Chris Dumez.

As Navigation is a LocalDOMWindowProperty it always has
access to its own ScriptExecutionContext.

* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::navigation):
* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::Navigation):
(WebCore::Navigation::initializeEntries):
(WebCore::Navigation::scriptExecutionContext const):
(WebCore::Navigation::serializeState):
(WebCore::Navigation::navigate):
(WebCore::Navigation::updateCurrentEntry):
(WebCore::Navigation::updateForNavigation):
* Source/WebCore/page/Navigation.h:
* Source/WebCore/page/Navigation.idl:

Canonical link: https://commits.webkit.org/276478@main
